File: CalcErroArredondamento.py
import sympy as sp
import math
class CalcErroArredondamento:
    def __init__(self):
        self.equation = ""        # equação de entrada do usuário
        self.expr = ""            # expressão matemática convertida
        self.f = ""               # uso da expressão como função
        self.intervaloStart = 0.0 # intervalo inicial
        self.intervaloEnd = 0.0   # intervalo final
        self.trapezios = 0        # quantidade de trapézios
        self.casasDecimais = 0    # casas decimais desejada
        self.altura = 0           # usada para calcular altura
        self.fatorAltura = 0      # altura dos trapézios
        self.calculo = 0          # somatório das áreas do trapézio
        self.tabela = ""          # tabela com os valores de x dentro do intervalo citado e com a função
        self.tabelaCriada = False # validação se a tabela foi ou não criada
        self.calcErro = False     # validação se o erro de arredondamento foi ou não calculado
        self.receberDadosdoProblema() # função para usuário inserir os dados de entrada
        self.calculoTrap()            # função para calcular a altura dos trapézio
        self.menu()                   # menu de opções para o usuário
    
    def createEqua(self, expressao: str):
        # está função converte qualquer string em expressão matemática para ser usada no código
        x = sp.Symbol('x')
        expr = sp.sympify(expressao)  # Interpreta a string como expressão simbólica
        f = sp.lambdify(x, expr, modules='numpy')
        return f, expr

    def receberDadosdoProblema(self):
        # usuário deve digitar os dados de entrada desejado
        while True:
            try:
                self.equation = input("Digite a função:")
                self.f, self.expr = self.createEqua(self.equation)
                break
            except:
                print("ERRO! DIGITE A FUNÇÃO CORRETAMENTE!")
        while True:
            try:
                self.intervaloStart = float(input("Digite o valor inicial do intervalo:"))
                break
            except:
                print("ERRO! DIGITE APENAS NÚMEROS!")
        while True:
            try:
                self.intervaloEnd = float(input("Digite o valor final do intervalo:"))
                break
            except:
                print("ERRO! DIGITE APENAS NÚMEROS!")
        while True:
            try:
                self.trapezios = int(input("Digite a quantidade de trapézios:"))
                break
            except:
                print("ERRO! DIGITE APENAS NÚMEROS!")
        while True:
            try:
                self.casasDecimais = int(input("Digite a quantidade de casas decimais desejada para as respostas"))
                break
            except:
                print("ERRO! DIGITE APENAS NÚMEROS!")
    
    def calculoTrap(self):
        # função para calcular a altura dos trapézio
        self.altura = self.intervaloEnd - self.intervaloStart
        self.fatorAltura = self.altura / self.trapezios
        print(f"A altura de cada Trapézio (eixo X) consiste na razão entre a diferença \ndos valores do intervalo e o número de trapézios declarado!")
    
    def criarTabela(self):
        # criação da tabela com os dados dentro do limite do intervalo digitado pelo usuário e seus respectivos valores da função
        i = 0
        aux = []
        equat = []
        soma = 0
        aux.append(self.intervaloStart) # insere valor inicial do intervalo

        construirTabela = ""
        for i in range(0, self.trapezios+1): # percorre todos os trapézios dentro do intervalo
            equat.append(self.f(aux[i]))     # adiciona o valor de f(x) na lista
            if i < self.trapezios:
                aux.append(aux[i] + self.fatorAltura) # adiciona o fator em x exemplo: x = 2 --> x = 2.5 assim por diante
        # equat[0] = equat[0] / 2 
        # equat[self.trapezios] = equat[self.trapezios] /2
        construirTabela = "  X              f(X) \n"
        for i in range(0, self.trapezios+1): # construção da tabela visual para o usuário
            # soma += raizQuad[i]
            soma += equat[i]
            construirTabela += f"  {aux[i]:.{self.casasDecimais}f}       {equat[i]:.{self.casasDecimais}f}\n"
        construirTabela += "-----------------------\n"
        construirTabela += f"Soma = {soma:.{self.casasDecimais}f}"
        self.tabela = construirTabela
        self.tabelaCriada = True
    
    def exibirTabela(self):
        if not self.tabelaCriada:
            print("ERRO! A TABELA AINDA NÃO FOI CRIADA!")
            return
        print(self.tabela)
    
    def calcErroArred(self):
        self.calculo = 0
        aux = []
        aux.append(self.intervaloStart) # adiciona inicio do intervalo em aux
        soma = 0
        equat = []
        for i in range(0, self.trapezios+1): # percorre os trapézios e adicionar o valor de f(x) em equat[]
            equat.append(self.f(aux[i]))
            if(i < self.trapezios):
                aux.append(aux[i] + self.fatorAltura) # adiciona o fator em x exemplo: x = 2 --> x = 2.5 assim por diante
        
        for i in range(0, self.trapezios): # somatório das áreas dos trapézios
            self.calculo += ((equat[i+1] + equat[i]) / 2) * self.fatorAltura
            
        temp = round(self.calculo * math.pow(10, self.casasDecimais))
        ultimaCasa = int(temp % 10)
        vlArred = 0.5 * math.pow(10, -self.casasDecimais) # valor de arredondamento para calcular o erro
        erroArredondamento = self.trapezios * vlArred * self.fatorAltura # erro de arredondamento
        self.calcErro = True
        print(f"|Ea| <= {self.trapezios} * {vlArred:.{self.casasDecimais+1}f} * {self.fatorAltura:.{self.casasDecimais}f} = {erroArredondamento:.{self.casasDecimais+1}f}")

    def getNumTrap(self):
        print(f"Número de Trapézios = {self.trapezios}")
    def getCasasDecimais(self):
        print(f"Número de Casas Decimais = {self.casasDecimais}")
    def getAlturaCadaTrap(self):
        print(f"Altura de cada Trapézio = {self.fatorAltura}")
    def getInicioIntervalo(self):
        print(f"Valor de início do intervalo = {self.intervaloStart}")
    def getFimIntervalo(self):
        print(f"Valor do fim do intervalo = {self.intervaloEnd}")
    def getValorSomatorio(self):
        if not self.calcErro:
            print(f"ERRO! A SOMATÓRIA AINDA NÃO FOI CALCULADA!")
            return
        print(f"Resultado do Somatório = {self.calculo:.{self.casasDecimais}f}")

    def menu(self):
        while True:
            try:
                escolha = int(input("=== MENU ===\n" +
                                "1 - Calcula Erro de Arredondamento\n" +
                                "2 - Criar Tabela de 'x' e 'f(x)'\n" +
                                "3 - Exibir Tabela\n" +
                                "4 - Dados do Problema\n" +
                                "0 - Sair\n\n" +
                                "Escolha uma opção:\n"))
                if escolha == 0:
                    break
                elif escolha == 1:
                    self.calcErroArred()
                elif escolha == 2:
                    self.criarTabela()
                    print("Tabela de valores criada e preenchida com sucesso!")
                elif escolha == 3:
                    self.exibirTabela()
                elif escolha == 4:
                    self.getInicioIntervalo()
                    self.getFimIntervalo()
                    self.getNumTrap()
                    self.getCasasDecimais()
                    self.getAlturaCadaTrap()
                    self.getValorSomatorio()

            except:
                print("Erro, selecione uma opção corretamente")

CalcErroArredondamento()

    

File: CalcErroArredondamento.java
import javax.swing.JOptionPane;

public class CalcErroArredondamento {

    private double vlInicial;
    private double vlFinal;
    private int qtdCasasDec;
    private int numTrap;
    private double area;
    private double altura;
    private double fatorAltura;
    private double calculo;
    private boolean tabelaCriada = false;
    private boolean calcErro = false;
    private String tabela;
    private String entrada;
    
    public CalcErroArredondamento() {
        receberDadosdoProblema();
        calculoTrap();
        menu();
    }

    public void receberDadosdoProblema() {
        try {
            entrada = JOptionPane.showInputDialog("Digite o valor inicial do intervalo:");

            if (entrada == null) {
                JOptionPane.showMessageDialog(null, "PROGRAMA CANCELADO!");
                System.exit(0);
            }

            vlInicial = Double.parseDouble(entrada);

        }catch(NumberFormatException e){
            JOptionPane.showMessageDialog(null, "ERRO! DIGITE APENAS NÚMEROS!");
            System.exit(0);
        }
        try {
            entrada = JOptionPane.showInputDialog("Digite o valor final do intervalo:");

            if (entrada == null) {
                JOptionPane.showMessageDialog(null, "PROGRAMA CANCELADO!");
                System.exit(0);
            }

            vlFinal = Double.parseDouble(entrada);

        }catch(NumberFormatException e){
            JOptionPane.showMessageDialog(null, "ERRO! DIGITE APENAS NÚMEROS!");
            System.exit(0);
        }
        try {
            entrada = JOptionPane.showInputDialog("Digite a quantidade de trapézios:");

            if (entrada == null) {
                JOptionPane.showMessageDialog(null, "PROGRAMA CANCELADO!");
                System.exit(0);
            }

            numTrap = Integer.parseInt(entrada);

        }catch(NumberFormatException e){
            JOptionPane.showMessageDialog(null, "ERRO! DIGITE APENAS NÚMEROS!");
            System.exit(0);
        }
        try {
            entrada = JOptionPane.showInputDialog("Digite a quantidade de casas decimais desejada para as respostas");

            if (entrada == null) {
                JOptionPane.showMessageDialog(null, "PROGRAMA CANCELADO!");
                System.exit(0);
            }

            qtdCasasDec = Integer.parseInt(entrada);

        }catch(NumberFormatException e){
            JOptionPane.showMessageDialog(null, "ERRO! DIGITE APENAS NÚMEROS!");
            System.exit(0);
        }
    }
    
    public void calculoTrap() {
        altura = vlFinal - vlInicial;
        fatorAltura = altura / numTrap;
        JOptionPane.showMessageDialog(null,"A altura de cada Trapézio (eixo X) consiste na razão entre a diferença \ndos valores do intervalo e o número de trapézios declarado!");
    }

    public void criarTabela() {
        int i;
        double[] aux = new double[numTrap + 1];
        double[] raizQuad = new double[numTrap + 1];
        double soma;
        aux[0] = vlInicial;
        soma = 0;
        StringBuilder construirTabela = new StringBuilder();
        for(i=0;i<=numTrap;i++) {
            raizQuad[i] = Math.sqrt(aux[i]);
            if (i < numTrap) { aux[i + 1] = aux[i] + fatorAltura; } // Evita estouro
        }
        raizQuad[0] = raizQuad[0] / 2;
        raizQuad[numTrap] = raizQuad[numTrap] / 2;
        construirTabela.append("   X          f(X)\n");
        for(i=0;i<=numTrap;i++) {
            soma += raizQuad[i];
            construirTabela.append(String.format("%." + qtdCasasDec + "f    %." + qtdCasasDec + "f\n", aux[i], raizQuad[i]));
        }
        construirTabela.append("-------------------\n");
        construirTabela.append(String.format("            %." + qtdCasasDec + "f", soma));
        tabela = construirTabela.toString();
        tabelaCriada = true;
    }
    
    public void exibirTabela() {
        if (!tabelaCriada) {
            JOptionPane.showMessageDialog(null, "ERRO! A TABELA AINDA NÃO FOI CRIADA!");
            return;
        }
        JOptionPane.showMessageDialog(null, tabela);
    }
    
    public void calcErroArred() {
        int i, soma, ultimaCasa;
        double[] aux = new double[numTrap + 1];
        double[] raizQuad = new double[numTrap + 1];
        double vlArred, erroArredondamento;
        long temp;
        aux[0] = vlInicial;
        soma = 0;
        for(i=0;i<=numTrap;i++) {
            raizQuad[i] = Math.sqrt(aux[i]);
            if (i < numTrap) { aux[i + 1] = aux[i] + fatorAltura; }
        }
        for(i=0;i<numTrap;i++){
            calculo += ((raizQuad[i+1] + raizQuad[i]) / 2) * fatorAltura;
        }
        temp = Math.round(calculo * Math.pow(10, qtdCasasDec));
        ultimaCasa = (int)(temp % 10);
        vlArred = 0.5 * Math.pow(10, -qtdCasasDec);
        erroArredondamento = numTrap * vlArred * fatorAltura;
        calcErro = true;
        JOptionPane.showMessageDialog(null,String.format("|Ea| <= %d * %f * %f = %f", numTrap, vlArred, fatorAltura, erroArredondamento)); 
    }

    public void getNumTrap() {
        JOptionPane.showMessageDialog(null,String.format("Número de Trapézios = %d", numTrap)); 
    }
    
    public void getCasasDecimais() {
        JOptionPane.showMessageDialog(null,String.format("Número de Casas Decimais = %d", qtdCasasDec)); 
    }
    
    public void getAlturaCadaTrap() {
        JOptionPane.showMessageDialog(null,String.format("Altura de cada Trapézio = %f", fatorAltura)); 
    }

    public void getInicioIntervalo() {
        JOptionPane.showMessageDialog(null,String.format("Valor de início do intervalo = %f", vlInicial)); 
    }
    
    public void getFimIntervalo() {
        JOptionPane.showMessageDialog(null,String.format("Valor do fim do intervalo = %f", vlFinal)); 
    }
    
    public void getValorSomatorio() {
        if (!calcErro) {
            JOptionPane.showMessageDialog(null, "ERRO! A SOMATORIA AINDA NÃO FOI CALCULADA!");
            return;
        }
        JOptionPane.showMessageDialog(null,String.format("Resultado do Somatorio = %." + qtdCasasDec + "f", calculo)); 
    }
    
    public void menu() {
        int opcao = -1;
        do {
            try {
                
                entrada = JOptionPane.showInputDialog(
                "=== MENU ===\n" +
                "1 - Calcula Erro de Arredondamento\n" +
                "2 - Criar Tabela de 'x' e 'f(x)'\n" +
                "3 - Exibir Tabela\n" +
                "4 - Dados do Problema\n" +
                "0 - Sair\n\n" +
                "Escolha uma opção:"
                );
            
                if(entrada == null) {
                    JOptionPane.showMessageDialog(null, "PROGRAMA FINALIZADO!");
                    System.exit(0);
                }

                opcao = Integer.parseInt(entrada);

                switch(opcao) {
                    case 1 -> calcErroArred();
                    case 2 -> { 
                        criarTabela();
                        JOptionPane.showMessageDialog(null, "Tabela de valores criada e preenchida com sucesso!");
                    }
                    case 3 -> exibirTabela();
                    case 4 -> {
                        getInicioIntervalo();
                        getFimIntervalo();
                        getNumTrap();
                        getCasasDecimais();
                        getAlturaCadaTrap();
                        getValorSomatorio();
                    }
                    case 0 -> JOptionPane.showMessageDialog(null, "SAINDO DO PROGRAMA...");
                    default -> JOptionPane.showMessageDialog(null, "OPCAO INVALIDA!");
                }
            }catch(NumberFormatException e) {
                JOptionPane.showMessageDialog(null, "ERRO! DIGITE APENAS NÚMEROS!");
                System.exit(0);
            }
        } while(opcao != 0);
    }
    
    public static void main(String[] args) {
        new CalcErroArredondamento();
    }
}